packages/common/infra/src/workspace/engine/doc/index.ts

```diff
@@ -0,0 +1,181 @@
+  remotePart: DocEngineRemotePart | null;
+
+  storage: DocStorageInner;
+  eventBus: DocEventBusInner;
+
+  engineState = LiveData.computed(get => {
+    const localState = get(this.localPart.engineState);
+    if (this.remotePart) {
+      const remoteState = get(this.remotePart?.engineState);
+      return {
+        total: remoteState.total,
+        syncing: remoteState.syncing,
+        saving: localState.syncing,
+        retrying: remoteState.retrying,
+        errorMessage: remoteState.errorMessage,
+      };
+    }
+    return {
+      total: localState.total,
+      syncing: localState.syncing,
+      saving: localState.syncing,
+      retrying: false,
+      errorMessage: null,
+    };
+  });
+
+  docState(docId: string) {
+    const localState = this.localPart.docState(docId);
+    const remoteState = this.remotePart?.docState(docId);
+    return LiveData.computed(get => {
+      const local = get(localState);
+      const remote = remoteState ? get(remoteState) : null;
+      return {
+        ready: local.ready,
+        saving: local.syncing,
+        syncing: local.syncing || remote?.syncing,
+      };
+    });
+  }
+
+  constructor(
+    storage: DocStorage,
+    eventBus: DocEventBus,
+    private readonly server?: DocServer | null
+  ) {
+    const clientId = nanoid();
+    this.storage = new DocStorageInner(storage);
+    this.eventBus = new DocEventBusInner(eventBus);
+    this.localPart = new DocEngineLocalPart(
+      clientId,
+      this.storage,
+      this.eventBus
+    );
+    this.remotePart = this.server
+      ? new DocEngineRemotePart(
+          clientId,
+          this.storage,
+          this.server,
+          this.eventBus
+        )
+      : null;
+  }
+
+  abort = new AbortController();
+
+  start() {
+    this.abort.abort(MANUALLY_STOP);
+    this.abort = new AbortController();
+    Promise.all([
+      this.localPart.mainLoop(this.abort.signal),
+      this.remotePart?.mainLoop(this.abort.signal),
+    ]).catch(err => {
+      if (err === MANUALLY_STOP) {
+        return;
+      }
+      logger.error('Doc engine error', err);
+    });
+    return this;
+  }
+
+  stop() {
+    this.abort.abort(MANUALLY_STOP);
+  }
+
+  async resetSyncStatus() {
+    this.stop();
+    await this.storage.clearSyncMetadata();
+    await this.storage.clearServerClock();
+  }
+
+  addDoc(doc: YDoc, withSubDocs = true) {
+    this.localPart.actions.addDoc(doc);
+    this.remotePart?.actions.addDoc(doc.guid);
+
+    if (withSubDocs) {
+      const subdocs = doc.getSubdocs();
+      for (const subdoc of subdocs) {
+        this.addDoc(subdoc, false);
+      }
+      doc.on('subdocs', ({ added }: { added: Set<YDoc> }) => {
+        for (const subdoc of added) {
+          this.addDoc(subdoc, false);
+        }
+      });
+    }
+  }
+
+  setPriority(docId: string, priority: number) {
+    this.localPart.setPriority(docId, priority);
+    this.remotePart?.setPriority(docId, priority);
+  }
+
+  waitForSaved() {
```

Comment:
Wondering how to distinguish the difference between `saved` / `synced` / `ready`? Are they implementation details, or should be sensible for engine users?

Reply:
* saved: ydoc changes have been saved to storage, and the browser can be safely closed without losing data.

* synced: is fully synchronized with the server.

* ready: is a strange flag, which means that the doc has been loaded and the data can be modified. (is not force, you can still modify it if you know you are creating some new data)
  
  This is a temporary solution to deal with the yjs overwrite issue.
  
  if content is loaded from storage
or if content is pulled from the server, it will be true, otherwise be false.
  
  For example, when opening a doc that is not in storage, ready = false until the content is pulled from the server.